### ISSUE TITLE
[XLA?]  jnp.log1p for complex inputs looses accuracy

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#7004]
I think the complex implementation of `jax.numpy.log1p` has issues.
Could someone point me to where it's implemented? I find it hard to go through the tensorflow repository to find what i'm looking for...

Anyhow, with the latest releases of jax and jaxlib:

```python
>>> import jax
>>> float(jax.numpy.log(1.00001))   # real-valued, log, inaccurate
1.0013530300057027e-05
>>> float(jax.numpy.log1p(0.00001))  # real-valued, log1p, accurate
9.999949725170154e-06
>>> complex(jax.numpy.log(1.00001+0.0j)) # complxe-valued, log, inaccurate
(1.0013530300057027e-05+0j)
>>> complex(jax.numpy.log1p(0.00001+0.0j)) # complxe-valued, log1p, also inaccurate
(1.001348027784843e-05+0j)
```

I would expect the same result for both real and complex valued cases.

For comparison, numpy does the correct thing:
```python
>>> np.log1p(0.00001)
9.999950000333332e-06
>>> np.log1p(0.00001+0.0j)
(9.999950000398841e-06+0j)
```

I could not find the numpy implementation, however, in their repo. I will paste here the Julia implementation, which also gives correct results, and checks if the real part is finite and in that case uses a smart trick to increase accuracy.
```julia
function log1p(z::Complex{T}) where T
    zr,zi = reim(z)
    if isfinite(zr)
        isinf(zi) && return log(z)
        # This is based on a well-known trick for log1p of real z,
        # allegedly due to Kahan, only modified to handle real(u) <= 0
        # differently to avoid inaccuracy near z==-2 and for correct branch cut
        u = one(float(T)) + z
        u == 1 ? convert(typeof(u), z) : real(u) <= 0 ? log(u) : log(u)*z/(u-1)
    elseif isnan(zr)
        Complex(zr, zr)
    elseif isfinite(zi)
        Complex(T(Inf), copysign(zr > 0 ? zero(T) : convert(T, pi), zi))
    else
        Complex(T(Inf), T(NaN))
    end
end
```

**Comment by hawkinsp** [comment]
The implementation XLA is using on CPU and GPU is here:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/elemental_ir_emitter.cc;l=572

It looks like the Julia implementation is certainly simpler.

**Comment by PhilipVinc** [comment]
Thanks for the link!

Looking at the implementation, `log1p(a+bi) = .5*log((a+1)^2+b^2) + i*atan2(b, a + 1)` where a is the real part and b is the imaginary part, it seems that when `b=0` this reduces to `log1p(a+0i) = .5*log((1+a)^2)` which is the non-optimized formula.

I'm not trying to push for the julia implementation, which might have issues due to the if-else statements, but the current XLA implementation has an issue so it may make sense to change it?

Should I move this issue to tensorflow? Can I push to escalate this issue somehow?

**Comment by hawkinsp** [comment]
Let me have a go at implementing the Julia version.

**Comment by hawkinsp** [comment]
By the way, I'm not sure classic NumPy *is* doing the right thing here:
```
In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.20.3'

In [3]: np.log1p(np.complex64(1e-5))
Out[3]: (1.001353e-05+0j)

In [4]: np.log1p(np.complex128(1e-5))
Out[4]: (9.999950000398841e-06+0j)
```

If I'm reading the right code, it's pretty similar to the current XLA one:
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/umath/funcs.inc.src#L329

**Comment by hawkinsp** [comment]
We came up with a simpler fix that seems to work, at least in this case:
```
 log((a+1)+bi) = .5*log((a+1)^2+b^2) + i*atan2(b, a+1)
 log((a+1)+bi) = .5*log(a*a + 2*a + 1 + b*b) + i*atan2(b, a+1)
log((a+1)+bi) = .5*log1p(a*a + 2*a + b*b) + i*atan2(b, a+1)
```

It has the additional advantage of being robust to reassociation; the `z + 1 - 1` currently gets constant-folded in some cases to `z + (1 - 1)`.

**Comment by PhilipVinc** [comment]
This looks great, and increases precision because you are now using the real-valued log1p...

Thanks!

**Comment by hawkinsp** [comment]
This is now fixed at jaxlib head. Hope that helps!

**Comment by WarrenWeckesser** [comment]
> This is now fixed at jaxlib head. Hope that helps!

Can anyone point me to the current implementation of `log1p`?  I'm reviewing a PR to update NumPy's implementation, and I'd like to refer to the Jax version.

By the way, a problem with the formula `log((a+1)+bi) = .5*log1p(a*a + 2*a + b*b) + i*atan2(b, a+1)` is that a call such as `log1p(1e200 + 1e200j)` (with double precision floating point) will result in overflow in the expressions `a*a` and `b*b`.
